### PR TITLE
opentelemetry-collector: 0.40.0 -> 0.43.1, opentelemetry-collector-contrib: init at 0.43.0

### PR DESCRIPTION
**Pull request by 06kellyjac**
- opentelemetry-collector: 0.40.0 -> 0.43.1
- opentelemetry-collector-contrib: init at 0.43.0

###### Motivation for this change

* separately package `opentelemetry-collector` and `opentelemetry-collector-contrib`
* update each to their lates releases: `0.43.1` and `0.43.0` respectively
* dropped vendor proxy since it seems to build fine
* added ldflags for a smaller build (trims 3MB off the contrib build)
* dropped `CGO_ENABLED = 0;` since it builds fine without
  * we can add this back in if it's important these packages are statically linked or people can build it themselves adding `CGO_ENABLED = 0;` via an overlay
* added myself as a maintainer

the core build is 7 times smaller than the contrib build

otelcontribcol:

```bash
λ ls -al ./result/bin/otelcontribcol
.r-xr-xr-x root root 118 MB Thu Jan  1 01:00:01 1970  ./result/bin/otelcontribcol

λ du -h ./result/bin/otelcontribcol
119M	./result/bin/otelcontribcol

# the Makefile builds with the vX.X.X prefix
λ ./result/bin/otelcontribcol -v
otelcontribcol version v0.43.0
```

otelcorecol:

```bash
λ ls -al ./result/bin/otelcorecol
.r-xr-xr-x root root 17 MB Thu Jan  1 01:00:01 1970  ./result/bin/otelcorecol

λ du -h ./result/bin/otelcorecol
17M	./result/bin/otelcorecol

# the cmd/otelcorecol/main.go file does not have a vX.X.X prefix
λ ./result/bin/otelcorecol -v
otelcorecol version 0.43.1
```

I've added release notes to cover the breaking change and added a script incase people try to call `otelcontribcol` using the the non contrib package:

```bash
λ ./result/bin/otelcontribcol
ERROR: otelcontribcol is now in `pkgs.opentelemetry-collector-contrib`, call the collector with `otelcorecol` or move to `pkgs.opentelemetry-collector-contrib`
```

cc: @uri-canva what are your thoughts

###### Things done

- Built on platform(s)
  - [X] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [X] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [X] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [X] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Review by uri-canva — CHANGES_REQUESTED** [review]
The fixed output derivation for the deps doesn't match between darwin and linux, I'm not sure if this is intended to happen because the dependencies might have os or architecture specific differences, or if they should have been stripped. Other than that it looks fine.

File: pkgs/tools/misc/opentelemetry-collector/contrib.nix [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+  pname = "opentelemetry-collector-contrib";
+  version = "0.43.0";
+
+  src = fetchFromGitHub {
+    owner = "open-telemetry";
+    repo = "opentelemetry-collector-contrib";
+    rev = "v${version}";
+    sha256 = "sha256-ktzP+ugG2sa0v8B1Zp47o8Bmpxv98zQyFyWf9QfQRoQ=";
+  };
+  vendorSha256 = "sha256-E4P2aNE31ATYjvEOnAOsbIRLInfROUc7Gv1eIrjXVZM=";
```
Comment (zowoq) [inline pkgs/tools/misc/opentelemetry-collector/contrib.nix:16]: Need to use `proxyVendor` here like in the original derivation and update the `vendorSha256` to fix the darwin/linux hash mismatch.

**Comment by uri-canva** [comment]
From what I can tell none of the other packages have os specific vendor hashes, and `proxyVendor` was specifically made to work around filesystem case insensitivity issue causing hashes for macOS and linux to differ, so it's expected to be the same.
I guess this is why I had to use `proxyVendor` before, though I had forgotten about it.

File: pkgs/tools/misc/opentelemetry-collector/contrib.nix (hunk 1 above)
Comment (uri-canva) [inline pkgs/tools/misc/opentelemetry-collector/contrib.nix:16]:
```suggestion
  vendorSha256 = "sha256-0E52YSWlq1ebHA3kR9Qo/6ufug9R+z1cSD9AfbN/Mi0=";
  proxyVendor = true;
```

**Review by uri-canva — CHANGES_REQUESTED** [review]

File: pkgs/tools/misc/opentelemetry-collector/contrib.nix (hunk 1 above)
Comment (06kellyjac) [reply pkgs/tools/misc/opentelemetry-collector/contrib.nix:16]: ah, ok I'll add it back with a note

Comment (06kellyjac) [reply pkgs/tools/misc/opentelemetry-collector/contrib.nix:16]: ah, ok I'll add it back with a note

**Review by uri-canva — APPROVED** [review]

**Comment by uri-canva** [comment]
Looks all good to me, but the PR is still in draft, did you mean to do something else before merging it?

**Comment by 06kellyjac** [comment]
Yeah I just wanted to double check you and other's thoughts on moving the contrib version and if the release note and little message script were enough.

Also was there a reason the build was CGO_ENABLE 0 before? Just to get some library to build or was it to get a static binary to copy out to non-nix systems?
If thats the most typical usecase we can always put it back

**Comment by uri-canva** [comment]
I think I copied it from upstream: https://github.com/open-telemetry/opentelemetry-collector/blob/5ee1fec9fe877a180b30bd4f29979ca21a2fa335/Makefile#L196, so maybe it's best to add it to both derivations.

**Comment by 06kellyjac** [comment]
Yeah for upstream the goal is to make the binary portable to any system. Within nixpkgs we have more control of whether or not someone has available the dynamic libraries we link to.

It seems to be preferred to lean towards linking rather than static binaries in nixpkgs. It's easy enough for someone to add an overlay to build it statically, or if enough users request it change back in issues/matrix we can always switch it back at that point

**Comment by uri-canva** [comment]
Sounds good!

**Pull request merged**